File: mesh/generic/xbeeRadio.py
from mesh.generic.radio import Radio, RadioMode
import mesh.generic.gpio as GPIO

class XbeeRadio(Radio):
    def __init__(self, serial, config, sleepPin = []):
        Radio.__init__(self, serial, config)
        self.sleepPin = sleepPin
        # Setup sleep pin
        if self.sleepPin:
            GPIO.setup(self.sleepPin, "out")
            self.setOff() # default to off mode
        pass

    def setSleep(self):
        if self.sleepPin:
            GPIO.output(self.sleepPin, "high")
        self.mode = RadioMode.sleep

    def setOff(self):
        self.setSleep()
        self.mode = RadioMode.off

    def setReceive(self):
        self.wake()
        self.mode = RadioMode.receive

    def setTransmit(self):
        self.wake()
        self.mode = RadioMode.transmit

    def wake(self):
        if self.sleepPin:
            GPIO.output(self.sleepPin, "low")

File: cpp/comm/xbeeRadio.cpp
#include "comm/xbeeRadio.hpp"
#include "GPIOWrapper/GPIOWrapper.hpp"

namespace comm {
    XbeeRadio::XbeeRadio(serial::Serial * serialIn, RadioConfig & configIn, int sleepPinIn) :
        Radio(serialIn, configIn),
        sleepPin(sleepPinIn)
    {
        if (sleepPinIn > -1) { // sleep pin available
            GPIOWrapper::setupPin((unsigned int)sleepPinIn, GPIOWrapper::OUTPUT);
        }
    }
    
    bool XbeeRadio::setOff(void) {
        setSleep();
        mode = RADIO_OFF;
        return true;
    }

    bool XbeeRadio::setSleep(void) {
        if (sleepPin > -1) {
            GPIOWrapper::setValue((unsigned int)sleepPin, GPIOWrapper::HIGH);
        }

        mode = RADIO_SLEEP;

        return true;
    }

    bool XbeeRadio::setReceive(void) {
        wake();
        mode = RADIO_RECEIVE;
        return true;
    }

    bool XbeeRadio::setTransmit(void) {
        wake();
        mode = RADIO_TRANSMIT;
        return true;
    }

    void XbeeRadio::wake(void) {
        if (sleepPin > -1) {
            GPIOWrapper::setValue((unsigned int)sleepPin, GPIOWrapper::LOW);
        }
    }
}
